### PR TITLE
Misc fixes

### PR DESCRIPTION
**Pull request by FreeMasen**
This PR addresses a number of bug and/or ergonomic issues with the library.

First and foremost this begins the process of centralizing a lot of the code that was duplicated in both the request and response modules, this makes addressing the following issues easier.

- the `iter` and `serialize` APIs were only available for the "builder" path, which is a bit of an ergonomic nightmare (#3), this fixes that
- The `get_content_length` method wasn't available on both types (one was left not updated previously from the unprefixed setter)
- The `get_*`, `fill_*` and `seralize*` methods are all nearly identical

Additionally, this PR addresses some general cleanup around docs test and type checking

### CODE REVIEW

**Comment by dljsjr** [comment]
I haven't had a chance to dig in really deep yet, does this refactor address chunked transfer encoding for socket ltn12 sources?

**Comment by FreeMasen** [comment]
> I haven't had a chance to dig in really deep yet, does this refactor address chunked transfer encoding for socket ltn12 sources?

Yes, the primary logic for this is [here](https://github.com/cosock/luncheon/blob/e956f08b52778fe14346a6eee39f75e42d2c5475/luncheon/shared.lua#L83-L205), with additional modifications [here](https://github.com/cosock/luncheon/blob/e956f08b52778fe14346a6eee39f75e42d2c5475/luncheon/shared.lua#L304-L310) for the `iter` implementation

**Comment by cjswedes** [comment]
@FreeMasen what are you thinking that you still want to do here? Currently there are some failing tests relating to the chunked encoding. Let me know if there is anything I can take on to help get this completed.

**Comment by FreeMasen** [comment]
> @FreeMasen what are you thinking that you still want to do here? Currently there are some failing tests relating to the chunked encoding. Let me know if there is anything I can take on to help get this completed.

It looks like the CI run passed and I am passing locally, are you seeing failures locally?

**Comment by cjswedes** [comment]
I think I had something off with my environment. I am seeing them pass now.

File: luncheon/shared.lua [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+    or (self._send_state.sent or 0) < #self.body do
+    local s, e = SharedLogic.send_body_chunk(self)
+    if not s then
+      return nil, e
+    end
+  end
+  return 1
+end
+
+return {
```
Comment (cjswedes) [inline luncheon/shared.lua]: It seems like this entire shared logic module could be included into the Request/Response classes as a metatable. We dont have to make such a change right now, but I was curious if going down the route of setting up inheritance in this way was considered.

**Review by cjswedes — APPROVED** [review]

File: luncheon/shared.lua (hunk 1 above)
Comment (FreeMasen) [reply luncheon/shared.lua]: yeah, this comes down to me not really understanding lua's inheritance model and this was the easiest way to share that logic. I am not opposed to using a meta-table to achieve the same thing but just haven't had the energy to tackle both the parsing logic and the inheritance layout.

Comment (cjswedes) [reply luncheon/shared.lua]: https://github.com/cosock/luncheon/pull/10

**Pull request merged**